services/core/shared/core/blocks.js

```diff
@@ -0,0 +1,207 @@
+	'generatorAddress',
+	'generatorUsername',
+	'height',
+	'numberOfTransactions',
+	'previousBlockId',
+	'unixTimestamp',
+	'totalAmount',
+	'totalFee',
+	'isFinal',
+	['height', 'isFinal'],
+	['generatorPublicKey', 'numberOfTransactions'],
+	['generatorAddress', 'numberOfTransactions'],
+	['generatorUsername', 'numberOfTransactions'],
+	['generatorPublicKey', 'totalAmount'],
+	['generatorAddress', 'totalAmount'],
+	['generatorUsername', 'totalAmount'],
+	['generatorPublicKey', 'unixTimestamp'],
+	['generatorAddress', 'unixTimestamp'],
+	['generatorUsername', 'unixTimestamp'],
+];
+
+let lastBlock = {};
+
+const getSelector = (params) => {
+	const result = {};
+	const selector = {};
+	if (params.height) selector.height = Number(params.height);
+	if (params.blockId) selector.id = String(params.blockId);
+	if (params.fromTimestamp) selector.unixTimestamp = { $gte: Number(params.fromTimestamp) };
+	if (params.toTimestamp) selector.unixTimestamp = { $lte: Number(params.toTimestamp) };
+	if (params.generatorPublicKey) selector.generatorPublicKey = String(params.generatorPublicKey);
+
+	if (params.limit) result.limit = Number(params.limit);
+	if (Number(params.offset) >= 0) result.skip = Number(params.offset);
+
+	// Not supported by the API but useful for caching
+	if (params.numberOfTransactions) selector.numberOfTransactions = {
+		$gte: Number(params.numberOfTransactions),
+	};
+	if (params.isFinal) selector.isFinal = params.isFinal;
+	if (params.isImmutable) selector.isFinal = params.isImmutable;
+
+	result.selector = selector;
+
+	if (params.limit) result.limit = params.limit;
+	if (Number(params.offset) >= 0) result.skip = params.offset;
+
+	return result;
+};
+
+const pushToDb = async (blockDb, blocks) => {
+	const propList = [
+		'blockSignature',
+		'generatorAddress',
+		'generatorPublicKey',
+		'height',
+		'id',
+		'numberOfTransactions',
+		'payloadHash',
+		'payloadLength',
+		'previousBlockId',
+		'reward',
+		'totalAmount',
+		'totalFee',
+		'totalForged',
+		'unixTimestamp',
+		'version',
+		'isFinal',
+	];
+	const out = blocks.map(o => {
+		const obj = {};
+		propList.map(prop => obj[prop] = o[prop]);
+		return obj;
+	});
+	return blockDb.writeBatch(out);
+};
+
+const setLastBlock = block => lastBlock = block;
+const getLastBlock = () => lastBlock;
+
+const getBlocks = async (params = {}, skipCache = false) => {
+	const blockDb = await pouchdb(config.db.collections.blocks.name, indexList);
+
+	let blocks = {
+		data: [],
+	};
+
+	if (skipCache !== true && (params.blockId
+		|| params.numberOfTransactions
+		|| params.isFinal === true
+		|| params.isImmutable === true)) { // try to get from cache
+		const inputData = getSelector({
+			...params,
+			limit: params.limit || 10,
+			offset: params.offset || 0,
+		});
+		const dbResult = await blockDb.find(inputData);
+		if (dbResult.length > 0) {
+			blocks.data = dbResult.map((block) => ({
+				...block,
+				confirmations: (getLastBlock().height) - block.height,
```

Comment:
The last block always has a `confirmations` value of `1`. So, perhaps the `confirmations` property needs to be incremented by 1: `confirmations: (getLastBlock().height) - block.height + 1`

Reply:
Is it wrong? Consider this request: `https://mainnet.lisk.io/api/blocks?sort=height:desc&limit=1`. The Lisk Core reports the top block as confirmed once.

Reply:
Hypothetical scenario: If `block` happens to be the same as the one returned by `getLastBlock()`, then `getLastBlock().height - block.height` would result in `0`confirmations but the expected value is `1`, as reported from the Lisk Core.

Reply:
From the tests I've performed the number of confirmations is always 1.
Can you please try it out? 

```
$ http "https://testnet.lisk.io/api/blocks?sort=height:desc&limit=1" && http ":9901/api/v1/blocks?sort=height:desc&limit=1"
HTTP/1.1 200 OK
...
{
    "data": [
        {
            ...
            "confirmations": 1,
            ...
        }
    ],
    ...
}

HTTP/1.1 200 OK
...
{
    "data": [
        {
            ...
            "confirmations": 1,
            ...
        }
    ],
    ...
}
```

Reply:
With the `sort` and `limit` params in the request URL, the current implementation falls back to requesting data from the Lisk Core. The current code can be reached when we query for a specific block (identified by its `id` param) as below.

```
$ http "https://testnet.lisk.io/api/blocks?blockId=10393132071499841741" && http ":9901/api/v1/blocks?id=10393132071499841741"
HTTP/1.1 200 OK
access-control-allow-origin: *
content-security-policy: frame-ancestors 'none'
content-type: application/json
date: Wed, 28 Oct 2020 17:56:54 GMT
strict-transport-security: max-age=15768000
transfer-encoding: chunked
vary: Accept-Encoding
x-frame-options: DENY
x-ratelimit-limit: 0
x-ratelimit-remaining: 0

{
    "data": [
        {
            ...
            "confirmations": 1294,
            ...
        }
    ],
    ...
}

HTTP/1.1 200 OK
Connection: keep-alive
Content-Length: 704
Content-Type: application/json; charset=utf-8
Date: Wed, 28 Oct 2020 17:56:54 GMT
X-Request-ID: 65436d66-f50c-43f6-af1d-f007c8e24e95

{
    "data": [
        {
            ...
            "confirmations": 1293,
            ...
        }
    ],
    ...
}
```

Reply:
In fact, a more accurate calculation would be something like:

```
const lastBlock = getLastBlock();
...
     confirmations: lastBlock.height - block.height + lastBlock.confirmations
```